src/pybind/mgr/devicehealth/module.py

```diff
@@ -243,6 +255,66 @@ def show_device_metrics(self, devid, sample):
                 res[key] = v
         return (0, json.dumps(res, indent=4), '')
 
+    def life_expectancy_response(self):
+        mark_out_threshold_td = timedelta(seconds=int(self.mark_out_threshold))
+        warn_threshold_td = timedelta(seconds=int(self.warn_threshold))
+        health_warnings = []
+        devs = self.get("devices")
+        for dev in devs['devices']:
+            if 'life_expectancy_min' not in dev:
+                continue
+            # life_expectancy_(min/max) is in the format of:
+            # '%Y-%m-%d %H:%M:%S.%f', e.g.:
+            # '2019-01-20 21:12:12.000000'
+            life_expectancy_min = datetime.strptime(dev['life_expectancy_min'], '%Y-%m-%d %H:%M:%S.%f')
+            now = datetime.now()
+            if self.self_heal and life_expectancy_min - now <= mark_out_threshold_td:
+                if dev['daemons']:
+                    self.mark_out(dev['daemons'])
+                    # TODO: set_primary_affinity
+                else:
+                    self.log.error("Device %s is about to fail, but there are no OSDs associated with it to mark out." % dev['devid'])
+            # if we hit 'mark_out_threshold', health warning will be issued below:
+            if life_expectancy_min - now <= warn_threshold_td:
+                health_warnings.append('%s at %s:%s;'
+                                       ' Affected OSDs: %s;'
+                                       ' Life expectancy: between %s and %s'
+                                       % (dev['devid'],
+                                          dev['location'][0]['host'],
+                                          dev['location'][0]['dev'],
+                                          dev.get('daemons', 'none'),
+                                          dev['life_expectancy_min'],
+                                          dev.get('life_expectancy_max', 'unknown')))
+                # TODO: by default, dev['life_expectancy_max'] == '0.000000',
+                # so dev.get('life_expectancy_max', 'unknown')
+                # above should be altered. Let's decide on time range
+                # behaviours first.
+        if health_warnings:
+            self.set_health_checks({
+                'MGR_DEVICE_HEALTH': {
+                    'severity': 'warning',
+                    'summary': 'Imminent failure was detected on some devices.'
```

Comment:
Imminent failure anticipated for <num> device(s)

Reply:
fixed